### PR TITLE
Disallow generic type parameters from appearing within certain constants

### PR DESCRIPTION
**Pull request by BenLewis-Seequent**
This improves the diagonistics/fixes ICEs when generic parameters appear within some constants. For example of the new errors:
```rust
use std::mem::size_of;

fn foo<T>() {
    let _ = [0; size_of::<T>()]; // now errors with: type parameters can't appear within an array length expression
}

#[repr(usize)]
enum Enum<T> {
    V1 = size_of::<T>(), // now errors with: type parameters can't appear within an enum discriminant
}
``` 

Previously emitted confusing errors like:
```
error[E0277]: the size for values of type `T` cannot be known at compilation time
   --> src/main.rs:45:27
    |
35  | fn foo<T>() {
    |        - help: consider restricting this bound: `T: std::marker::Sized`
...
45  |     let _ = [0; size_of::<T>()];
    |                           ^ doesn't have a size known at compile-time
    |
    = help: the trait `std::marker::Sized` is not implemented for `T`
```
Note doesn't disallow referring to `Self` in either enum discriminants or array length expressions as there are cases where it can currently be sucessfully used, i.e.:
```rust
struct Foo;

impl Foo {
    const A: usize = 37;

    fn bar() -> [u8; Self::A]{
        [0; Self::A]
    }
}
```
I don't think that this disallows any more code that didn't either produce a compiler error before or caused an ICE.

cc #43408 
cc #67945 (doesn't actually fully fix it see: https://github.com/rust-lang/rust/issues/67945#issuecomment-575946966)

### CODE REVIEW

File: src/librustc_error_codes/error_codes/E0747.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+Any constant items, static items, array length expressions, or enum
```
Comment (Centril) [inline src/librustc_error_codes/error_codes/E0747.md]:
```suggestion
Any constant item, static item, array length expression, or enum
```

File: src/librustc_error_codes/error_codes/E0747.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+discriminants can't refer to type parameters.
```
Comment (Centril) [inline src/librustc_error_codes/error_codes/E0747.md]:
```suggestion
discriminant cannot refer to type parameters.
```

File: src/librustc_error_codes/error_codes/E0747.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+```
+use std::mem::size_of;
+
+fn foo<T>() {
+    let _ = vec![0; size_of::<T>()];
+}
+```
+
+Though enum discriminants can't refer to regular type parameters they can still
```
Comment (Centril) [inline src/librustc_error_codes/error_codes/E0747.md]:
```suggestion
While enum discriminants cannot refer to regular type parameters, they can still
```

File: src/librustc_error_codes/error_codes/E0747.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+```
+#[repr(u8)]
+enum Alpha {
+    V1 = 41,
+    V2 = Self::V1 as u8 + 1,
+}
+```
+
+Note, associated constants do not have this limitation and they can refer to
```
Comment (Centril) [inline src/librustc_error_codes/error_codes/E0747.md]:
```suggestion
Note that associated constants do not have this limitation and can refer to
```

File: src/librustc_error_codes/error_codes/E0747.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+struct Bar<T>(T);
+
+impl<T> Foo for Bar<T> {
+    const X: i32 = size_of::<T>();
+}
+```
+
+This is currently a limitation with the compiler and these restrictions may be
```
Comment (Centril) [inline src/librustc_error_codes/error_codes/E0747.md]:
```suggestion
This is currently a limitation with the compiler. These restrictions may be
```

File: src/librustc_resolve/diagnostics.rs [hunk 6]
```diff
@@ -360,6 +360,17 @@ impl<'a> Resolver<'a> {
                 err.span_label(span, "non-constant value");
                 err
             }
+            ResolutionError::GenericParamsInConst(source) => {
+                let mut err = struct_span_err!(
+                    self.session,
+                    span,
+                    E0447,
+                    "type parameters can't appear within {}",
```
Comment (Centril) [inline src/librustc_resolve/diagnostics.rs]:
```suggestion
                    "type parameters cannot appear within {}",
```

File: src/librustc_resolve/late.rs [hunk 7]
```diff
@@ -78,6 +78,36 @@ enum PatBoundCtx {
...
+    /// static item, e.g., `static _: u8 = X;`.
+    Static,
+    /// an array length expression, e.g., `[...; X]`.
+    ArrayLength,
+    /// an enum discriminant value, e.g., `enum Enum { V = X, }`.
+    EnumDiscriminant,
+    /// an associated constant in an impl or trait, e.g., `impl A { const _: u8 = X; }`.
+    AssocConstant,
+    /// a const generic argument, e.g., `Struct<{X}>`.
+    GenericArg,
```
Comment (Centril) [inline src/librustc_resolve/late.rs]: Capitalize the sentences in the variant docs.

File: src/librustc_resolve/late.rs [hunk 8]
```diff
@@ -382,10 +412,18 @@ impl<'a, 'tcx> Visitor<'tcx> for LateResolutionVisitor<'a, '_> {
     }
     fn visit_anon_const(&mut self, constant: &'tcx AnonConst) {
         debug!("visit_anon_const {:?}", constant);
-        self.with_constant_rib(|this| {
+        self.with_constant_rib(AnonConstUsage::ArrayLength, |this| {
```
Comment (Centril) [inline src/librustc_resolve/late.rs]: Feels odd that `visit_anon_const` always uses `::ArrayLength`.

File: src/librustc_resolve/lib.rs [hunk 9]
```diff
@@ -1,3 +1,4 @@
+// ignore-tidy-filelength
```
Comment (Centril) [inline src/librustc_resolve/lib.rs]: Can you split things in a follow up PR?

**Comment by Centril** [comment]
cc @petrochenkov @varkor

**Comment by eddyb** [comment]
I disagree with doing this in `rustc_resolve`, and the phrasing of the errors.
This is an implementation bug, they *should* be working (and they kind of do already with `#![feature(const_generics)]`, but that's a bit of a hack).

The reason I didn't want this in `rustc_resolve` is because I didn't want to risk introducing forwards-compatibility hazards - the behavior of `rustc_resolve` today is the one we want to keep, if we want to error we should probably do it in `rustc_typeck::astconv`, perhaps by checking any generic parameter `hir::Ty` we're converting to `Ty` actually exists in the generics *currently in scope* (as opposed to looking it up in the parameter's parent).

(there's probably one exception to this, and that would be the `enum` discriminant case, but we can allow name resolution on it and error later when trying to evaluate discriminants)

For improving errors, I would prefer doing a crater run with this changed to always be `Some`:
(now that we've special-cased literals, it might always work in practice)
https://github.com/rust-lang/rust/blob/6250d56355d72264ece721e8d0dc95b16a6824b1/src/librustc_typeck/collect.rs#L1022-L1033

**Pull request closed without merging**